# Data Analysis with Python - HW-1

#### <br>Data Source: https://stats.espncricinfo.com/ci/content/records/223646.html

### Q-1: Import the wickets.csv file in your jupyter notebook.


#### Importing required Libraries

In [1]:
import numpy as np
import pandas as pd

In [2]:
# reading a csv file

df = pd.read_excel("test_cricket.xlsx", sheet_name='wickets', engine="openpyxl")
print(df.shape)

(79, 14)


### Q-2: Display the first 10 rows of the dataframe

In [3]:
display(df.head(10))

Player       Span  Mat  Inns  Balls   Runs  Wkts  \
0  M Muralitharan (ICC/SL)  1992-2010  133   230  44039  18180   800   
1           SK Warne (AUS)  1992-2007  145   273  40705  17995   708   
2         A Kumble (INDIA)  1990-2008  132   236  40850  18355   619   
3        JM Anderson (ENG)  2003-2021  162   301  34791  16457   617   
4         GD McGrath (AUS)  1993-2007  124   243  29248  12186   563   
5          SCJ Broad (ENG)  2007-2021  148   272  29713  14502   523   
6            CA Walsh (WI)  1984-2001  132   242  30019  12688   519   
7            DW Steyn (SA)  2004-2019   93   171  18608  10077   439   
8      N Kapil Dev (INDIA)  1978-1994  131   227  27740  12867   434   
9        HMRKB Herath (SL)  1999-2018   93   170  25993  12157   433   

                   BBI                  BBM    Ave  Econ    SR   5  10  
0  1951-09-01 00:00:00               16/220  22.72  2.47  55.0  67  22  
1  1971-08-01 00:00:00               12/128  25.41  2.65  57.4  37  10  
2  1974-10-01 00:00:00               14/149  29.65  2.69  65.9  35   8  
3  1942-07-01 00:00:00  1971-11-01 00:00:00  26.67  2.83  56.3  30   3  
4  2021-08-24 00:00:00  2021-10-27 00:00:00  21.64  2.49  51.9  29   3  
5  2021-08-15 00:00:00               11/121  27.72  2.92  56.8  18   3  
6  1937-07-01 00:00:00                13/55  24.44  2.53  57.8  22   3  
7  1951-07-01 00:00:00  1960-11-01 00:00:00  22.95  3.24  42.3  26   5  
8  1983-09-01 00:00:00               11/146  29.64  2.78  63.9  23   2  
9                9/127               14/184  28.07  2.80  60.0  34   9

### Q-3: Create a markdown cell and explain the meaning of each column

In [4]:
# Printing the column names

print(df.columns)

Index(['Player',   'Span',    'Mat',   'Inns',  'Balls',   'Runs',   'Wkts',
          'BBI',    'BBM',    'Ave',   'Econ',     'SR',        5,       10],
      dtype='object')


##### Meaning of each column

> 1. **Player** - The name of the players/bowlers ordered as wickets taken from highest to lowest.
> 2. **Span** - Career span/duration of the players
> 3. **Mat** - Number of matches played
> 4. **Inns** - Total innings played by the player
> 5. **Balls** - Total number of balls bowled by the player
> 6. **Runs** - The number of runs conceded by the player
> 7. **Wkts** - Total number of wickets taken by the player
> 8. **BBI** - Best bowling in an innings of the player	
> 9. **BBM** - Best bowling in a match of the player
> 10. **Ave** - The average number of runs conceded per wicket by the player. (Runs/W)
> 11. **Econ** - The average number of runs conceded per over. (Runs/Overs bowled)
> 12. **SR** - The average number of balls bowled per wicket taken. (Balls/W)
> 13. **5** - The number of innings in which the bowler took at least five wickets.
> 14. **10** - The number of matches in which the bowler took at least ten wickets.

### Q-4: Find the number of rows and columns in the dataframe

In [5]:
# Number of rows
print("Number of Rows: ",df.shape[0] )

# Number of columns
print("Number of Columns: ",df.shape[1])

Number of Rows:  79
Number of Columns:  14


### Q-5: Find the data statistics and check for the data types

In [6]:
# Data Statistics

display(df.describe())

Mat        Inns         Balls          Runs        Wkts  \
count   79.000000   79.000000     79.000000     79.000000   79.000000   
mean    80.101266  144.797468  18630.303797   8595.506329  317.101266   
std     28.537692   51.042310   7190.036515   3080.256645  121.731587   
min     37.000000   67.000000   8785.000000   4846.000000  200.000000   
25%     60.500000  110.000000  13580.000000   6456.500000  229.000000   
50%     71.000000  129.000000  16498.000000   7742.000000  266.000000   
75%     93.000000  169.000000  21742.500000   9756.000000  374.500000   
max    166.000000  301.000000  44039.000000  18355.000000  800.000000   

             Ave       Econ         SR          5         10  
count  79.000000  79.000000  79.000000  79.000000  79.000000  
mean   27.466456   2.806582  59.187342  16.354430   2.797468  
std     3.657561   0.351666   9.349337   9.642372   3.235935  
min    20.940000   1.980000  41.200000   3.000000   0.000000  
25%    24.425000   2.600000  53.300000   9.500000   1.000000  
50%    28.000000   2.820000  57.400000  14.000000   2.000000  
75%    29.870000   3.080000  63.950000  20.500000   3.500000  
max    34.790000   3.460000  91.900000  67.000000  22.000000

In [7]:
# Data Statistics Transpose

display(df.describe().T)

count          mean          std      min        25%       50%  \
Mat     79.0     80.101266    28.537692    37.00     60.500     71.00   
Inns    79.0    144.797468    51.042310    67.00    110.000    129.00   
Balls   79.0  18630.303797  7190.036515  8785.00  13580.000  16498.00   
Runs    79.0   8595.506329  3080.256645  4846.00   6456.500   7742.00   
Wkts    79.0    317.101266   121.731587   200.00    229.000    266.00   
Ave     79.0     27.466456     3.657561    20.94     24.425     28.00   
Econ    79.0      2.806582     0.351666     1.98      2.600      2.82   
SR      79.0     59.187342     9.349337    41.20     53.300     57.40   
5       79.0     16.354430     9.642372     3.00      9.500     14.00   
10      79.0      2.797468     3.235935     0.00      1.000      2.00   

            75%       max  
Mat       93.00    166.00  
Inns     169.00    301.00  
Balls  21742.50  44039.00  
Runs    9756.00  18355.00  
Wkts     374.50    800.00  
Ave       29.87     34.79  
Econ       3.08      3.46  
SR        63.95     91.90  
5         20.50     67.00  
10         3.50     22.00

##### Statistical measures

- Count- total numbers of observation
- mean- average of the values
- std- standard deviation
- min- minimum value in the column
- 25%- 25% of all value is under the certain column value
- 50%- 50% of all value is under the certain column value
- 75%- 75% of all value is under the certain column value
- max- maximum value in the column


##### Observations

- On an average a player has played 80 matches
- Minimum 37 matches were played by a player
- 25% of players have played 60 matches or less
- 50% of players have played 71 matches or less
- 75% of players have played 93 matches or less
- Maximum 166 matches were played by a player
- Maxium 18355 runs were scored by a player
- Maximum 800 wickets were taken by a player
- Minimum 67 innings were played by a player

### Q-6:  Are there any missing values present in the dataset?

In [8]:
# checking for missing values and data types of each column

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Player  79 non-null     object 
 1   Span    79 non-null     object 
 2   Mat     79 non-null     int64  
 3   Inns    79 non-null     int64  
 4   Balls   79 non-null     int64  
 5   Runs    79 non-null     int64  
 6   Wkts    79 non-null     int64  
 7   BBI     79 non-null     object 
 8   BBM     79 non-null     object 
 9   Ave     79 non-null     float64
 10  Econ    79 non-null     float64
 11  SR      79 non-null     float64
 12  5       79 non-null     int64  
 13  10      79 non-null     int64  
dtypes: float64(3), int64(7), object(4)
memory usage: 8.8+ KB
None


##### Observations:

- Number of entries(rows) are 79 and and 14 columns.<br>
- There are-
> **3 Float** type variables- Bowling average, Economy rate, Strike rate<br>
> **4 String** type variables- Best Bowling (Innings), Best Bowling (Match), Player, Span<br>
> **7 Integer** type variables- Match, Innings, Balls, Runs, Wickets, 5 wickets (innings), 10 wickets (match)

- From the non-null columns, we can see that there are **no missing value**
---

### Q-7: Rename the column names appropriately

In [9]:
# Renamimng the column names to understand the Dataframe appropately

df = df.rename(columns={'Mat':'Match', 
                        'Inns':'Innings',
                        'Wkts': 'Wickets',
                        'BBI': 'Best_Bowling/Innings',
                        'BBM': 'Best_Bowling/Match',
                        'Ave': 'Average',
                        'Econ': 'Economy_Rate',
                        'SR': 'Strike_Rate',
                        5: '5_Wickets',
                        10: '10_Wickets'})

display(df.head())

Player       Span  Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan (ICC/SL)  1992-2010    133      230  44039  18180      800   
1           SK Warne (AUS)  1992-2007    145      273  40705  17995      708   
2         A Kumble (INDIA)  1990-2008    132      236  40850  18355      619   
3        JM Anderson (ENG)  2003-2021    162      301  34791  16457      617   
4         GD McGrath (AUS)  1993-2007    124      243  29248  12186      563   

  Best_Bowling/Innings   Best_Bowling/Match  Average  Economy_Rate  \
0  1951-09-01 00:00:00               16/220    22.72          2.47   
1  1971-08-01 00:00:00               12/128    25.41          2.65   
2  1974-10-01 00:00:00               14/149    29.65          2.69   
3  1942-07-01 00:00:00  1971-11-01 00:00:00    26.67          2.83   
4  2021-08-24 00:00:00  2021-10-27 00:00:00    21.64          2.49   

   Strike_Rate  5_Wickets  10_Wickets  
0         55.0         67          22  
1         57.4         37          10  
2         65.9         35           8  
3         56.3         30           3  
4         51.9         29           3

### Q-8: Remove a column from the dataframe

In [10]:
# Remove BBI and BBM columns as they have unorganized data.

df.drop(['Best_Bowling/Innings' , 'Best_Bowling/Match'], axis=1, inplace=True)

display(df.head())

Player       Span  Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan (ICC/SL)  1992-2010    133      230  44039  18180      800   
1           SK Warne (AUS)  1992-2007    145      273  40705  17995      708   
2         A Kumble (INDIA)  1990-2008    132      236  40850  18355      619   
3        JM Anderson (ENG)  2003-2021    162      301  34791  16457      617   
4         GD McGrath (AUS)  1993-2007    124      243  29248  12186      563   

   Average  Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  
0    22.72          2.47         55.0         67          22  
1    25.41          2.65         57.4         37          10  
2    29.65          2.69         65.9         35           8  
3    26.67          2.83         56.3         30           3  
4    21.64          2.49         51.9         29           3

### Removing & Splitting Player name and Country from Player column

In [11]:
#Removing the closing bracket

df['Player'] = df['Player'].str.replace(")"," ", regex=True)  

#splitting

df_country=df['Player'].str.split("(", expand=True)

In [12]:
display(df_country.head())

0        1
0  M Muralitharan   ICC/SL 
1        SK Warne      AUS 
2        A Kumble    INDIA 
3     JM Anderson      ENG 
4      GD McGrath      AUS

In [13]:
# Dropping the old player column

df.drop('Player', axis=1 ,inplace=True)  
display(df.head())

Span  Match  Innings  Balls   Runs  Wickets  Average  Economy_Rate  \
0  1992-2010    133      230  44039  18180      800    22.72          2.47   
1  1992-2007    145      273  40705  17995      708    25.41          2.65   
2  1990-2008    132      236  40850  18355      619    29.65          2.69   
3  2003-2021    162      301  34791  16457      617    26.67          2.83   
4  1993-2007    124      243  29248  12186      563    21.64          2.49   

   Strike_Rate  5_Wickets  10_Wickets  
0         55.0         67          22  
1         57.4         37          10  
2         65.9         35           8  
3         56.3         30           3  
4         51.9         29           3

#### Concatinating the new Dataframe & showing the new dataset with changes

In [14]:
# concatenating the new data set with the main data set
df = pd.concat([df, df_country], axis=1) 
#display(df.head())

# renaming new splitted column
df = df.rename(columns={0: 'Player',1: 'Country'})  
#display(df.head())

# new column sequence
column_sequence = ['Player' , 'Country' , 'Span' , 'Match' , 'Innings', 'Balls', 'Runs' , 'Wickets', 'Average' , 'Economy_Rate' , 'Strike_Rate' , '5_Wickets' , '10_Wickets']
df = df[column_sequence]


# Checking the new column sequence in the Data Set
display(df.head())

Player  Country       Span  Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan   ICC/SL   1992-2010    133      230  44039  18180      800   
1        SK Warne      AUS   1992-2007    145      273  40705  17995      708   
2        A Kumble    INDIA   1990-2008    132      236  40850  18355      619   
3     JM Anderson      ENG   2003-2021    162      301  34791  16457      617   
4      GD McGrath      AUS   1993-2007    124      243  29248  12186      563   

   Average  Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  
0    22.72          2.47         55.0         67          22  
1    25.41          2.65         57.4         37          10  
2    29.65          2.69         65.9         35           8  
3    26.67          2.83         56.3         30           3  
4    21.64          2.49         51.9         29           3

### Q: How many players played for ICC?

In [15]:
# Creating a function to check if played in ICC or not 

def ICC_status(x):
    if "ICC" in x:
        return "Yes"
    else:
        return "No"

# Applying function and creating new column for playing in ICC
df['ICC_status'] = df['Country'].apply(ICC_status)

# Removing "ICC/" from the country column
df['Country'] = df['Country'].str.replace("ICC/","", regex=True)
display(df.head(5))

Player Country       Span  Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan      SL   1992-2010    133      230  44039  18180      800   
1        SK Warne     AUS   1992-2007    145      273  40705  17995      708   
2        A Kumble   INDIA   1990-2008    132      236  40850  18355      619   
3     JM Anderson     ENG   2003-2021    162      301  34791  16457      617   
4      GD McGrath     AUS   1993-2007    124      243  29248  12186      563   

   Average  Economy_Rate  Strike_Rate  5_Wickets  10_Wickets ICC_status  
0    22.72          2.47         55.0         67          22        Yes  
1    25.41          2.65         57.4         37          10         No  
2    29.65          2.69         65.9         35           8         No  
3    26.67          2.83         56.3         30           3         No  
4    21.64          2.49         51.9         29           3         No

In [16]:
# Count number of total players who played for ICC

print("Number of total ICC ICC_Playing Players: ")
print(df['ICC_status'].value_counts())

Number of total ICC ICC_Playing Players: 
No     74
Yes     5
Name: ICC_status, dtype: int64


##### Observation

> *We see, there are only **5 ICC Playing players** in the data set, where **74 players didn't play for ICC**.*

### Q: How many different countries are present in this dataset?

In [17]:
print(df['Country'].value_counts())

AUS         18
ENG         13
INDIA       10
WI           9
SA           8
NZ           7
PAK          7
SL           3
ENG/ICC      2
BDESH        1
ZIM          1
Name: Country, dtype: int64


##### Observation

 > **11 different countries** are present in the dataset

### Q: How many Australian Bowlers are present in this dataset?

##### Answer:

> From the value counts, we see, there are  **18 Australian** bowlers but only **1 Bangladeshi** bowler.

### Q: Is there any Bangladeshi player present in this dataset?

##### Answer:

> Yes, Only **01** Bangladeshi player exist in the dataset. 

#### Reorganizing the dataframe after splitting start and end year

In [18]:
# splitting 'span' column
df_span = df['Span'].str.split("-", expand=True)

# concatenating the new dataframe with the main dataframe
df = pd.concat([df, df_span], axis=1)

# renaming new splitted column
df = df.rename(columns={0: "Start_Year",
                        1: "End_Year"})

# removing 'span' column
df = df.drop("Span", axis=1)

# new column sequence as a list
column_sequence = ['Player', 'Country','Start_Year', 'End_Year', 'Match' , 'Innings', 'Balls', 'Runs' , 'Wickets', 'Average' , 'Economy_Rate' , 'Strike_Rate' , '5_Wickets' , '10_Wickets', 'ICC_status']

# implementing the new column sequence
df = df[column_sequence]

#### Changing Start and End Year from object type to integer

In [19]:
df['End_Year']=pd.to_numeric(df['End_Year'])
df['Start_Year']=pd.to_numeric(df['Start_Year'])

In [20]:
# total years of career of the players

df['Career_Span']=df['End_Year']-df['Start_Year']

# rearranging columns
column_sequence = ['Player', 'Country','Start_Year', 'End_Year', 'Career_Span' , 'Match', 'Innings', 'Balls', 'Runs', 'Wickets' , 'Average' , 'Economy_Rate' , 'Strike_Rate' , '5_Wickets' , '10_Wickets', 'ICC_status']
df = df[column_sequence] 
display(df.head())

Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
0  M Muralitharan      SL         1992      2010           18    133      230   
1        SK Warne     AUS         1992      2007           15    145      273   
2        A Kumble   INDIA         1990      2008           18    132      236   
3     JM Anderson     ENG         2003      2021           18    162      301   
4      GD McGrath     AUS         1993      2007           14    124      243   

   Balls   Runs  Wickets  Average  Economy_Rate  Strike_Rate  5_Wickets  \
0  44039  18180      800    22.72          2.47         55.0         67   
1  40705  17995      708    25.41          2.65         57.4         37   
2  40850  18355      619    29.65          2.69         65.9         35   
3  34791  16457      617    26.67          2.83         56.3         30   
4  29248  12186      563    21.64          2.49         51.9         29   

   10_Wickets ICC_status  
0          22        Yes  
1          10         No  
2           8         No  
3           3         No  
4           3         No

### Which player(s) had played for the longest period of time?

In [21]:
display(df.sort_values(by='Career_Span', ascending = False).head(3))

Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
21    Imran Khan     PAK         1971      1992           21     88      142   
55     GS Sobers      WI         1954      1974           20     93      159   
9   HMRKB Herath      SL         1999      2018           19     93      170   

    Balls   Runs  Wickets  Average  Economy_Rate  Strike_Rate  5_Wickets  \
21  19458   8258      362    22.81          2.54         53.7         23   
55  21599   7999      235    34.03          2.22         91.9          6   
9   25993  12157      433    28.07          2.80         60.0         34   

    10_Wickets ICC_status  
21           6         No  
55           0         No  
9            9         No

> So, we see, **Imran Khan** has longest career span of 21 years. And 2nd and third longest are following by GS Sobers (20) and HMRKB Herath (19)

### Q: Which player(s) had played for the shortest period of time?

In [22]:
display(df.sort_values(by='Career_Span', ascending = True).head(3))

Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
44    GP Swann     ENG         2008      2013            5     60      109   
71    K Rabada      SA         2015      2021            6     47       86   
54  Yasir Shah     PAK         2014      2021            7     45       83   

    Balls  Runs  Wickets  Average  Economy_Rate  Strike_Rate  5_Wickets  \
44  15349  7642      255    29.96          2.98         60.1         17   
71   8785  4846      213    22.75          3.30         41.2         10   
54  13601  7247      235    30.83          3.19         57.8         16   

    10_Wickets ICC_status  
44           3         No  
71           4         No  
54           3         No

> **GP Swann** has shortest career span of 5 years. And 2nd and third shortest are following by K Rabada (6) and Yasir Shah (7)

### Q: Which player had the lowest economy rate?

In [23]:
display(df.sort_values(by='Economy_Rate', ascending = True).head(1))

Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
32  LR Gibbs      WI         1958      1976           18     79      148   

    Balls  Runs  Wickets  Average  Economy_Rate  Strike_Rate  5_Wickets  \
32  27115  8989      309    29.09          1.98         87.7         18   

    10_Wickets ICC_status  
32           2         No

##### Answer:

> **LR Gibbs** has the lowest economy rate of 1.98

### Q: Which player had the lowest strike rate?

In [24]:
display(df.sort_values(by='Strike_Rate', ascending = True).head(1))

Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
71  K Rabada      SA         2015      2021            6     47       86   

    Balls  Runs  Wickets  Average  Economy_Rate  Strike_Rate  5_Wickets  \
71   8785  4846      213    22.75           3.3         41.2         10   

    10_Wickets ICC_status  
71           4         No

##### Answer:

> **K Rabada** has the lowest strike rate of 41.2

### Q: Which player had the lowest bowling average?

In [25]:
display(df.sort_values(by='Average', ascending = True).head(1))

Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
19  MD Marshall      WI         1978      1991           13     81      151   

    Balls  Runs  Wickets  Average  Economy_Rate  Strike_Rate  5_Wickets  \
19  17584  7876      376    20.94          2.68         46.7         22   

    10_Wickets ICC_status  
19           4         No

##### Answer:

> **MD Marshall** has the lowest bowling average of 20.94